### PR TITLE
TSK-1965: create endpoints for all state transitions

### PR DESCRIPTION
**Pull request by norman-schmidt**
---
Release Notes:

```
**New**
* 4 new REST endpoints:
API_V1 + "tasks/{taskId}/claim/force"
API_V1 + "tasks/{taskId}/complete/force"
API_V1 + "tasks/{taskId}/terminate"
API_V1 + "tasks/{taskId}/{isRead}

**Breaking Changes**
* until now the REST endpoint to delete a Task: DELETE Request on API_V1 + tasks/{taskId} has used the forceDelete method of the taskService --> this is now split into two endpoints:
DELETE Request on API_V1 + tasks/{taskId} to delete a task normaly
DELETE Request on API_V1 + tasks/{taskId}/force to force delete a task


```
 
### For the submitter:
- [ ] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- [x] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [x] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
https://sonarcloud.io/summary/new_code?id=norman-schmidt_taskana&branch=TSK-1965
- [ ] After integration of the PR, I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [ ] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I put the ticket in review
- [ ] After integration of the pull request, I verified our [bluemix test environment](http://taskana.mybluemix.net/taskana) is not broken

### Verified by the reviewer:
- [x] Commit message format → TSK-XXX: Your commit message.
- [x] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [x] SonarCloud analysis meets our standards
- [ ] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [ ] PR fulfills the ticket
- [ ] Edge cases and unwanted side effects are tested
- [ ] Readability

### CODE REVIEW

File: rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java [hunk 1]
```diff
@@ -198,8 +198,31 @@ public ResponseEntity<TaskRepresentationModel> claimTask(
...
+  }
+
+  /**
+   * This endpoint force claims a Task if possible even if it is already claimed by someone else.
+   *
+   * @param taskId the Id of the Task which should be force claimed
+   * @param userName TODO: this is currently not used
+   * @return the force claimed Task
+   * @throws TaskNotFoundException if the requested Task does not exist.
+   * @throws InvalidStateException if the state of the requested Task is not READY.
```
Comment (mustaphazorgati) [inline rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java:209]: if the state of Task with taskId is in {@linkplain TaskState.END_STATES}. This needs to be adjusted for the Java API aswell.

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java [hunk 2]
```diff
@@ -1306,6 +1302,62 @@ void should_CancelTask_when_CallingCancelEndpoint() {
...
+
+    // claim
+    String url2 =
+        restHelper.toUrl(
+            RestEndpoints.URL_TASKS_ID_CLAIM, "TKI:000000000000000000000000000000000033");
+    ResponseEntity<TaskRepresentationModel> claimResponse =
+        TEMPLATE.exchange(url2, HttpMethod.POST, auth, TASK_MODEL_TYPE);
+
+    assertThat(claimResponse.getBody()).isNotNull();
+    assertThat(claimResponse.getStatusCode().is2xxSuccessful()).isTrue();
```
Comment (mustaphazorgati) [inline rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java]: We typically verify the exact status code value. What do u think is better?

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java [hunk 3]
```diff
@@ -1306,6 +1302,62 @@ void should_CancelTask_when_CallingCancelEndpoint() {
...
+
+    // claim
+    String url2 =
+        restHelper.toUrl(
+            RestEndpoints.URL_TASKS_ID_CLAIM_FORCE, "TKI:000000000000000000000000000000000029");
+    ResponseEntity<TaskRepresentationModel> claimResponse =
+        TEMPLATE.exchange(url2, HttpMethod.POST, auth, TASK_MODEL_TYPE);
+
+    assertThat(claimResponse.getBody()).isNotNull();
+    assertThat(claimResponse.getStatusCode().is2xxSuccessful()).isTrue();
```
Comment (mustaphazorgati) [inline rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java]: We typically verify the exact status code value. What do u think is better?

File: rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java [hunk 4]
```diff
@@ -405,6 +405,29 @@ public ResponseEntity<TaskRepresentationModel> completeTask(@PathVariable String
...
+  }
+
+  /**
+   * This endpoint force completes a Task.
+   *
+   * @param taskId Id of the requested Task to force complete.
+   * @return the force completed Task
+   * @throws TaskNotFoundException if the requested Task does not exist.
+   * @throws InvalidOwnerException cannot be thrown.
+   * @throws InvalidStateException if Task wasn't claimed previously.
```
Comment (mustaphazorgati) [inline rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java:417]: if the state of the Task with taskId is TERMINATED or CANCELED (link both)

Comment (mustaphazorgati) [reply rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java:417]: Needs also adjustment in java api

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java [hunk 5]
```diff
@@ -1558,6 +1558,62 @@ void should_ForceRequestChanges_When_CurrentUserIsNotTheOwner() {
...
+        restHelper.toUrl(
+            RestEndpoints.URL_TASKS_ID_COMPLETE, "TKI:000000000000000000000000000000000102");
+    ResponseEntity<TaskRepresentationModel> requestedChangesResponse =
+        TEMPLATE.exchange(url2, HttpMethod.POST, auth, TASK_MODEL_TYPE);
+
+    assertThat(requestedChangesResponse.getBody()).isNotNull();
+    assertThat(requestedChangesResponse.getStatusCode()).isEqualTo(HttpStatus.OK);
+    repModel = requestedChangesResponse.getBody();
+    assertThat(repModel.getOwner()).isEqualTo("user-1-2");
+    assertThat(repModel.getState()).isEqualTo(TaskState.COMPLETED);
```
Comment (mustaphazorgati) [inline rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java]: please adjust the instance variables + comment to match with the complete operation

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java [hunk 6]
```diff
@@ -1558,6 +1558,62 @@ void should_ForceRequestChanges_When_CurrentUserIsNotTheOwner() {
...
+        restHelper.toUrl(
+            RestEndpoints.URL_TASKS_ID_COMPLETE_FORCE, "TKI:000000000000000000000000000000000028");
+    ResponseEntity<TaskRepresentationModel> requestedChangesResponse =
+        TEMPLATE.exchange(url2, HttpMethod.POST, auth, TASK_MODEL_TYPE);
+
+    assertThat(requestedChangesResponse.getBody()).isNotNull();
+    assertThat(requestedChangesResponse.getStatusCode()).isEqualTo(HttpStatus.OK);
+    repModel = requestedChangesResponse.getBody();
+    assertThat(repModel.getOwner()).isEqualTo("user-1-2");
+    assertThat(repModel.getState()).isEqualTo(TaskState.COMPLETED);
```
Comment (mustaphazorgati) [inline rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java]: please adjust the instance variables + comment to match with the complete operation

File: rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java [hunk 7]
```diff
@@ -73,6 +73,43 @@ public class TaskController {
...
+    Task fromResource = taskRepresentationModelAssembler.toEntityModel(taskRepresentationModel);
+    Task createdTask = taskService.createTask(fromResource);
+
+    return ResponseEntity.status(HttpStatus.CREATED)
+        .body(taskRepresentationModelAssembler.toModel(createdTask));
+  }
+
+  // endregion
+
+  // region GET
```
Comment (mustaphazorgati) [inline rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java:111]: GET -> READ?

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java [hunk 8]
```diff
@@ -1726,6 +1726,34 @@ void should_CancelTask_when_CallingCancelEndpoint() {
     }
   }
 
+  @Nested
+  @TestInstance(Lifecycle.PER_CLASS)
+  class TerminateTasks {
+    @Test
+    void should_TerminateTask_when_CallingCancelEndpoint() {
```
Comment (mustaphazorgati) [inline rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java:1733]: should_TerminateTask is enough

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java [hunk 9]
```diff
@@ -1726,6 +1726,34 @@ void should_CancelTask_when_CallingCancelEndpoint() {
...
+      assertThat(theTaskRepresentationModel.getState()).isEqualTo(TaskState.CLAIMED);
+
+      // cancel the task
+      String url2 =
+          restHelper.toUrl(
+              RestEndpoints.URL_TASKS_ID_TERMINATE, "TKI:000000000000000000000000000000000103");
+      responseGet = TEMPLATE.exchange(url2, HttpMethod.POST, auth, TASK_MODEL_TYPE);
+
+      assertThat(responseGet.getBody()).isNotNull();
+      assertThat(responseGet.getBody().getState()).isEqualTo(TaskState.TERMINATED);
```
Comment (mustaphazorgati) [inline rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java:1753]: please adjust the instance variables + comment to match with the terminate operation

File: rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java [hunk 10]
```diff
@@ -539,7 +539,27 @@ public ResponseEntity<TaskRepresentationModel> updateTask(
...
+   * @return the updated Task
+   * @throws TaskNotFoundException if the requested Task does not exist.
+   * @throws NotAuthorizedException if the current user has no read permission for the Workbasket
+   *     the Task is in
+   * @title Set a Task read or unread
+   */
+  @PostMapping(path = RestEndpoints.URL_TASKS_ID_SET_READ)
+  @Transactional(rollbackFor = Exception.class)
+  public ResponseEntity<TaskRepresentationModel> setTaskRead(
+      @PathVariable String taskId, @PathVariable boolean isRead)
```
Comment (mustaphazorgati) [inline rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java:556]: I prever the `isRead` variable to be passed via request body.

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java [hunk 11]
```diff
@@ -1904,4 +1904,64 @@ void should_ThrowException_When_CancelClaimingOfClaimedTaskByAnotherUser() {
...
+      String url2 =
+          restHelper.toUrl(
+              RestEndpoints.URL_TASKS_ID_SET_READ,
+              "TKI:000000000000000000000000000000000025",
+              true);
+      ResponseEntity<TaskRepresentationModel> setReadResponse =
+          TEMPLATE.exchange(url2, HttpMethod.POST, auth, TASK_MODEL_TYPE);
+
+      assertThat(setReadResponse.getBody()).isNotNull();
+      assertThat(setReadResponse.getStatusCode().is2xxSuccessful()).isTrue();
```
Comment (mustaphazorgati) [inline rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java:1934]: Please verify status code by itself

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java [hunk 12]
```diff
@@ -1904,4 +1904,64 @@ void should_ThrowException_When_CancelClaimingOfClaimedTaskByAnotherUser() {
...
+      String url2 =
+          restHelper.toUrl(
+              RestEndpoints.URL_TASKS_ID_SET_READ,
+              "TKI:000000000000000000000000000000000027",
+              false);
+      ResponseEntity<TaskRepresentationModel> setReadResponse =
+          TEMPLATE.exchange(url2, HttpMethod.POST, auth, TASK_MODEL_TYPE);
+
+      assertThat(setReadResponse.getBody()).isNotNull();
+      assertThat(setReadResponse.getStatusCode().is2xxSuccessful()).isTrue();
```
Comment (mustaphazorgati) [inline rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java:1962]: please verify status code by itself

**Review by mustaphazorgati — CHANGES_REQUESTED** [review]
I am missing a forceDeleteTask REST endpoint,

Other than that it looks very good! :)

File: rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java (hunk 1 above)
Comment (norman-schmidt) [reply rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java:209]: It seems like we don't link in the REST Docs, I'll write it plain text. In the Java API I'll link of course.

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java (hunk 8 above)
Comment (norman-schmidt) [reply rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java:1733]: will also adjust for `should_CancelTask`

File: rest/taskana-rest-spring/src/main/java/pro/taskana/common/rest/RestEndpoints.java [hunk 13]
```diff
@@ -59,7 +60,7 @@ public final class RestEndpoints {
   public static final String URL_TASKS_ID_TERMINATE = API_V1 + "tasks/{taskId}/terminate";
   public static final String URL_TASKS_ID_TRANSFER_WORKBASKET_ID =
       API_V1 + "tasks/{taskId}/transfer/{workbasketId}";
-  public static final String URL_TASKS_ID_SET_READ = API_V1 + "tasks/{taskId}/{isRead}";
+  public static final String URL_TASKS_ID_SET_READ = API_V1 + "tasks/{taskId}/set-readF";
```
Comment (mustaphazorgati) [inline rest/taskana-rest-spring/src/main/java/pro/taskana/common/rest/RestEndpoints.java:63]: Is this a typo?

**Review by mustaphazorgati — CHANGES_REQUESTED** [review]

**Review by mustaphazorgati — APPROVED** [review]

**Pull request merged**